File: src/services/dropbox_service.py
import dropbox
import os
from pathlib import Path

class DropboxService:
    def __init__(self):
        self.dropbox = None
        self.model_folder = '/gift_question_models'
        
    async def initialize(self):
        try:
            if not os.getenv('DROPBOX_ACCESS_TOKEN'):
                raise ValueError('DROPBOX_ACCESS_TOKEN nie je nastavený v .env súbore')

            self.dropbox = dropbox.Dropbox(os.getenv('DROPBOX_ACCESS_TOKEN'))
            await self.ensure_model_folder()
            print('Dropbox služba inicializovaná')
        except Exception as e:
            print(f'Chyba pri inicializácii Dropbox: {str(e)}')
            raise

    async def ensure_model_folder(self):
        try:
            try:
                self.dropbox.files_get_metadata(self.model_folder)
            except dropbox.exceptions.ApiError:
                self.dropbox.files_create_folder(self.model_folder)
        except Exception as e:
            print(f'Chyba pri vytváraní priečinka: {str(e)}')
            raise

    async def upload_model(self, model_path):
        try:
            model_files = ['model.json', 'weights.bin']
            for file_name in model_files:
                file_path = Path(model_path) / file_name
                dropbox_path = f"{self.model_folder}/{file_name}"
                
                with open(file_path, 'rb') as f:
                    self.dropbox.files_upload(f.read(), dropbox_path, mode=dropbox.files.WriteMode.overwrite)

            print('Model bol úspešne nahratý na Dropbox')
            return True
        except Exception as e:
            print(f'Chyba pri nahrávaní modelu: {str(e)}')
            return False

    async def download_model(self, model_path):
        try:
            model_files = ['model.json', 'weights.bin']
            all_files_downloaded = True

            for file_name in model_files:
                dropbox_path = f"{self.model_folder}/{file_name}"
                local_path = Path(model_path) / file_name

                try:
                    _, res = self.dropbox.files_download(dropbox_path)
                    with open(local_path, 'wb') as f:
                        f.write(res.content)
                except dropbox.exceptions.ApiError:
                    all_files_downloaded = False
                    break

            if all_files_downloaded:
                print('Model bol úspešne stiahnutý z Dropboxu')
                return True
            else:
                print('Niektoré súbory modelu neboli nájdené na Dropboxe')
                return False
        except Exception as e:
            print(f'Chyba pri sťahovaní modelu: {str(e)}')
            return False 

File: src/services/dropboxService.js
const { Dropbox } = require('dropbox');
const fs = require('fs');
const path = require('path');

class DropboxService {
    constructor() {
        this.dropbox = null;
        this.modelFolder = '/gift_question_models';
    }

    async initialize() {
        try {
            if (!process.env.DROPBOX_ACCESS_TOKEN) {
                throw new Error('DROPBOX_ACCESS_TOKEN nie je nastavený v .env súbore');
            }

            this.dropbox = new Dropbox({
                accessToken: process.env.DROPBOX_ACCESS_TOKEN
            });

            // Vytvoríme priečinok pre modely ak neexistuje
            await this.ensureModelFolder();
            
            console.log('Dropbox služba inicializovaná');
        } catch (error) {
            console.error('Chyba pri inicializácii Dropbox:', error);
            throw error;
        }
    }

    async ensureModelFolder() {
        try {
            try {
                await this.dropbox.filesGetMetadata({ path: this.modelFolder });
            } catch (error) {
                if (error.status === 409) { // Folder už existuje
                    return;
                }
                // Vytvoríme nový priečinok
                await this.dropbox.filesCreateFolder({ path: this.modelFolder });
            }
        } catch (error) {
            console.error('Chyba pri vytváraní priečinka:', error);
            throw error;
        }
    }

    async uploadModel(modelPath) {
        try {
            const modelFiles = ['model.json', 'weights.bin'];
            const uploadPromises = modelFiles.map(async (fileName) => {
                const filePath = path.join(modelPath, fileName);
                const dropboxPath = `${this.modelFolder}/${fileName}`;

                // Načítame súbor
                const fileContent = fs.readFileSync(filePath);
                
                // Nahrajeme na Dropbox
                await this.dropbox.filesUpload({
                    path: dropboxPath,
                    contents: fileContent,
                    mode: { '.tag': 'overwrite' }
                });
            });

            await Promise.all(uploadPromises);
            console.log('Model bol úspešne nahratý na Dropbox');
            return true;
        } catch (error) {
            console.error('Chyba pri nahrávaní modelu:', error);
            return false;
        }
    }

    async downloadModel(modelPath) {
        try {
            const modelFiles = ['model.json', 'weights.bin'];
            const downloadPromises = modelFiles.map(async (fileName) => {
                const dropboxPath = `${this.modelFolder}/${fileName}`;
                const localPath = path.join(modelPath, fileName);

                try {
                    // Stiahneme súbor z Dropboxu
                    const { result } = await this.dropbox.filesDownload({ path: dropboxPath });
                    
                    // Uložíme lokálne
                    fs.writeFileSync(localPath, result.fileBinary);
                } catch (error) {
                    if (error.status === 409) { // Súbor neexistuje
                        return false;
                    }
                    throw error;
                }
            });

            const results = await Promise.all(downloadPromises);
            const allFilesDownloaded = results.every(result => result !== false);

            if (allFilesDownloaded) {
                console.log('Model bol úspešne stiahnutý z Dropboxu');
                return true;
            } else {
                console.log('Niektoré súbory modelu neboli nájdené na Dropboxe');
                return false;
            }
        } catch (error) {
            console.error('Chyba pri sťahovaní modelu:', error);
            return false;
        }
    }
}

module.exports = DropboxService; 